consensus/src/persistent_liveness_storage.rs

```diff
@@ -90,21 +96,120 @@ impl LedgerRecoveryData {
+        // LedgerInfo, we generate the virtual genesis block
+        let (latest_commit_id, latest_ledger_info_sig) =
+            if self.storage_ledger.ledger_info().ends_epoch() {
+                let genesis =
+                    Block::make_genesis_block_from_ledger_info(self.storage_ledger.ledger_info());
+                let genesis_qc = QuorumCert::certificate_for_genesis_from_ledger_info(
+                    self.storage_ledger.ledger_info(),
+                    genesis.id(),
+                );
+                let genesis_ledger_info = genesis_qc.ledger_info().clone();
+                let genesis_id = genesis.id();
+                blocks.push(genesis);
+                quorum_certs.push(genesis_qc);
+                (genesis_id, genesis_ledger_info)
+            } else {
+                (
+                    self.storage_ledger.ledger_info().consensus_block_id(),
+                    self.storage_ledger.clone(),
+                )
+            };
+
+        // sort by (epoch, round) to guarantee the topological order of parent <- child
+        blocks.sort_by_key(|b| (b.epoch(), b.round()));
+
+        let latest_commit_idx = blocks
+            .iter()
+            .position(|block| block.id() == latest_commit_id)
+            .ok_or_else(|| format_err!("unable to find root: {}", latest_commit_id))?;
+        let commit_block = blocks[latest_commit_idx].clone();
+        let root_quorum_cert = quorum_certs
+            .iter()
+            .find(|qc| qc.certified_block().id() == commit_block.id())
+            .ok_or_else(|| format_err!("No QC found for root: {}", commit_block.id()))?
+            .clone();
+
+        let (root_ordered_cert, root_commit_cert) = if order_vote_enabled {
+            // We are setting ordered_root same as commit_root. As every committed block is also ordered, this is fine.
+            // As the block store inserts all the fetched blocks and quorum certs and execute the blocks, the block store
+            // updates highest_ordered_cert accordingly.
+            let root_ordered_cert =
+                WrappedLedgerInfo::new(VoteData::dummy(), latest_ledger_info_sig.clone());
+            (root_ordered_cert.clone(), root_ordered_cert)
+        } else {
+            let root_ordered_cert = quorum_certs
+                .iter()
+                .find(|qc| qc.commit_info().id() == commit_block.id())
+                .ok_or_else(|| format_err!("No LI found for root: {}", latest_commit_id))?
+                .clone()
+                .into_wrapped_ledger_info();
+            let root_commit_cert = root_ordered_cert
+                .create_merged_with_executed_state(latest_ledger_info_sig)
+                .expect("Inconsistent commit proof and evaluation decision, cannot commit block");
+            (root_ordered_cert, root_commit_cert)
+        };
+
+        let window_start_round = blocks[latest_commit_idx]
+            .round()
+            .saturating_add(1)
```

Comment:
saturating_add seems unnecessary, something is probably wrong if we overflow u64